core/ov/source/ov_association.c

```diff
@@ -499,7 +502,34 @@ OV_DLLFNCEXPORT OV_RESULT ov_association_link(
 		*	in this case simply check, if the child link (anchor) is not already used
 		*/
 		if(Ov_Association_GetParent(passoc, pchild)) {
-			return OV_ERR_ALREADYEXISTS;
+			if(Ov_Association_GetParent(passoc, pchild)!=pparent)
+				return OV_ERR_ALREADYEXISTS;
+			/*
+			 * check if position is already correct
+			 */
+			switch(childhint){
+			case OV_PMH_DEFAULT:
+				// on placement and already linked
+				return OV_ERR_OK;
```

Comment:
If there is no link giving OV_PMH_DEFAULT would push the element to the end of the list.
If the link is already in the middle of the list the new code would keep it at that position.

So we have a different outcome in both scenarios.

Reply:
> So we have a different outcome in both scenarios.

I would not consider this an issue. IMHO, the semantics of "DEFAULT" are something like "place it whereever it suits best, I do not have a preference". So it'd be totally okay that the placement depends on some precondition like the presence of an existing link. And I personally find it quite intuitive that the default is the position of the already existing link, or at the end of the list if no link exists yet.

Reply:
You are the maintainer. So feel free to merge this.
But removing that small changed behavior would not make the code less elegant.

Reply:
A (known) breaking change should gain a profit. Simplicity of API (like your first point) is a valid point for that.
With the position of the element I do not see a profit.
But you are the maintainer. So feel free to merge this.

Reply:
As far as I understand, the new behaviour of the DEFAULT placement option is required to use the "relink" feature in a way that ensures the creation of a link, in case no link is present yet, but keeps an existing link untouched at its position in the list. If you explicitly want to create/move the link to the end of the list, you can use `OV_PMH_END`. Am I right, @LarsNt?

Reply:
In case I was not clear:
I would move the `case OV_PMH_DEFAULT:` three lines down and fall through into the `case OV_PMH_END:`.

Reply:
This would make `OV_PMH_DEFAULT` a pure alias of `OV_PMH_END`. So I would agree that default should express no specific placement preference.
Also your comparison is not quite fitting. Before this change the second call would have returned `KS_ERR_ALREADYEXISTS` and not changed anything in the order. Now it returns `KS_ERR_OK` and also not change the order.
So the change in this scenario is the return code from "already exists" to "ok". This was changed because it felt more important to distinguish between "the two objects are linked" vs "an existing link prevents linking" (which also returns `KS_ERR_ALREADYEXISTS`) than "a new link was created" vs "no new link was created".

Reply:
You are right. If you want to change this API not adjusting the order is probably ok.